File: scripts/update-uv.ts
import jszip from "jszip";
import * as tar from "tar";
import fs from "fs";
import zlib from "zlib";
import path from "path";
import { Readable } from "stream";
import { ReadableStream } from "stream/web";
import { pipeline } from "stream/promises";

const UV_LINK = "https://github.com/astral-sh/uv/releases/latest/download";
const FILES = [
  "uv-x86_64-unknown-linux-gnu.tar.gz",
  "uv-aarch64-unknown-linux-gnu.tar.gz",
  "uv-x86_64-apple-darwin.tar.gz",
  "uv-x86_64-pc-windows-msvc.zip",
  "uv-aarch64-pc-windows-msvc.zip",
];

// ファイルをダウンロードしつつ解凍(UV_LINK/FILE)
// リダイレクトは勝手に追跡してもらう
async function downloadAndDecompress(
  url: string,
  outputPath: string,
  decompressedDir: string
) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to download file: ${response.statusText}`);
  }
  if (!response.body) {
    throw new Error("Response body is null");
  }

  const ext = url.includes("windows") ? ".exe" : "";

  // zipとtar.gzで場合分け
  if (url.endsWith(".zip")) {
    const buffer = await response.arrayBuffer();
    const zip = await jszip.loadAsync(buffer);
    // uv(.exe)だけ抽出して保存
    const fileName = Object.keys(zip.files).find(
      (name) => name.endsWith("uv") || name.endsWith("uv.exe")
    );
    if (!fileName) {
      throw new Error("uv file not found in the zip archive");
    }
    const fileData = await zip.files[fileName].async("nodebuffer");
    fs.writeFileSync(decompressedDir + `/uv${ext}`, fileData);
  } else if (url.endsWith(".tar.gz")) {
    const responseStream = response.body;
    if (!responseStream) {
      throw new Error("Response body is null");
    }
    const gunzip = zlib.createGunzip();
    await pipeline(
      Readable.fromWeb(responseStream as unknown as ReadableStream<any>),
      gunzip,
      tar.extract({
        cwd: decompressedDir, // 出力先ディレクトリ
        filter: (path) => {
          // uv(.exe)だけ抽出
          return path.endsWith("uv") || path.endsWith("uv.exe");
        },
        strip: 1, // 最初のディレクトリを無視
      })
    ).catch((err) => {
      console.error("Pipeline failed:", err);
      throw err;
    });
  }

  // ファイルのos名のとおりにリネーム
  const urlName = path.basename(url).split(".")[0].slice(3); // 拡張子なしのファイル名(uv-を除く)
  const finalOutputPath = `${outputPath}/uv-${urlName}${ext}`;
  fs.renameSync(decompressedDir + `/uv${ext}`, finalOutputPath);
  console.log(`Downloaded and decompressed to ${finalOutputPath}`);
}

async function main() {
  const outputPath = `src-tauri/binaries/uv`;
  fs.mkdirSync(outputPath, { recursive: true });

  const decompressedDir = outputPath + "/tmp";
  if (!fs.existsSync(decompressedDir)) {
    fs.mkdirSync(decompressedDir, { recursive: true });
  }

  for (const file of FILES) {
    const fileUrl = `${UV_LINK}/${file}`;
    await downloadAndDecompress(fileUrl, outputPath, decompressedDir);
  }

  // 一時ディレクトリを削除
  fs.rmdirSync(decompressedDir, { recursive: true });
  console.log("All files downloaded and decompressed successfully.");
}

main().catch((error) => {
  console.error("Error:", error);
  process.exit(1);
});

File: scripts/update-uv.py
import os
import requests
import zipfile
import tarfile
import platform
import argparse
from io import BytesIO

UV_LINK = "https://github.com/astral-sh/uv/releases/latest/download"
FILES = [
    "uv-x86_64-unknown-linux-gnu.tar.gz",
    "uv-aarch64-unknown-linux-gnu.tar.gz",
    "uv-x86_64-apple-darwin.tar.gz",
    "uv-aarch64-apple-darwin.tar.gz",
    "uv-x86_64-pc-windows-msvc.zip",
    # "uv-aarch64-pc-windows-msvc.zip", # 現状aarch64のWindowsはサポートしないのでコメントアウト
]

parser = argparse.ArgumentParser()
parser.add_argument("-a", "--all", action="store_true", help="Download all files")
args = parser.parse_args()

def download_and_decompress(url, output_path, decompressed_dir):
    """
    ファイルをダウンロードして解凍する
    """
    print(f"Downloading {url}...")
    response = requests.get(url, allow_redirects=True, stream=True)
    response.raise_for_status()

    ext = ".exe" if "windows" in url else ""

    if url.endswith(".zip"):
        with zipfile.ZipFile(BytesIO(response.content)) as z:
            for member in z.infolist():
                if member.filename.endswith("uv") or member.filename.endswith("uv.exe"):
                    # uv(.exe)だけを抽出して保存
                    with open(os.path.join(decompressed_dir, f"uv{ext}"), "wb") as f:
                        f.write(z.read(member.filename))
                    break
    elif url.endswith(".tar.gz"):
        with tarfile.open(fileobj=BytesIO(response.content), mode="r:gz") as tar:
            for member in tar.getmembers():
                if member.name.endswith("uv"):
                    # uvだけを抽出して保存
                    extracted_file = tar.extractfile(member)
                    if extracted_file:
                        with open(os.path.join(decompressed_dir, "uv"), "wb") as f:
                            f.write(extracted_file.read())
                        break

    # ファイルのOS名のとおりにリネーム
    url_name = os.path.basename(url).split(".")[0][3:]  # 拡張子なしのファイル名(uv-を除く)
    final_output_path = os.path.join(output_path, f"uv-{url_name}{ext}")
    os.rename(os.path.join(decompressed_dir, f"uv{ext}"), final_output_path)
    print(f"Downloaded and decompressed to {final_output_path}")

def main():
    """
    メイン処理
    """
    output_path = os.path.join("src-tauri", "binaries", "uv")
    os.makedirs(output_path, exist_ok=True)

    decompressed_dir = os.path.join(output_path, "tmp")
    if not os.path.exists(decompressed_dir):
        os.makedirs(decompressed_dir, exist_ok=True)

    try:
        for file in FILES:
            if not platform.system().lower() in file and not args.all:
                print(f"Skipping {file} for current platform.")
                continue

            file_url = f"{UV_LINK}/{file}"
            download_and_decompress(file_url, output_path, decompressed_dir)

        print("All files downloaded and decompressed successfully.")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        # 一時ディレクトリを削除
        if os.path.exists(decompressed_dir):
            import shutil
            shutil.rmtree(decompressed_dir)

if __name__ == "__main__":
    main()